### ISSUE TITLE
Setting `irods::indexing::index` has no side effect when done from `msiModAVUMetadata`

### ISSUE THREAD

**Issue by dworkin** [irods/irods#6422]
## Bug Report

Setting the `irods::indexing::index` attribute only has the desired side effect when done through the imeta command (e.g. `imeta add -C $collection irods::indexing::index yoda::metadata elasticsearch`), not when it is done through `msiModAVUMetadata`.

### iRODS Version, OS and Version

iRODS 4.2.11, Centos 7.9

### What did you try to do?

```
msiModAVUMetadata("-C", *collection, "add", "irods::indexing::index", "yoda::metadata", "elasticsearch"); 
```

### Expected behavior

The delayed rule to update the search index is started.

### Observed behavior (including steps to reproduce, if applicable)

The attribute is set, but no delayed rule is started and the search index is not updated.

**Comment by korydraughn** [comment]
How did you execute the microservice?

Were you connected to the provider or consumer?

**Comment by dworkin** [comment]
The `msiModAVUMetadata` microservice was called from the iRODS rule set on the ICAT machine.

**Comment by korydraughn** [comment]
Sorry, I'll rephrase the questions.

Q. What client-side tool did you use?
Q. Is the client-side tool connecting to the provider or a consumer? _(this is very important in understanding why the microservice did not run)_
Q. What client-side command did you run?
Q. Are you calling `msiModAVUMetadata` from another PEP (e.g. `imkdir` -> `pep_api_coll_create`)?

Without knowing the details, I can say that if you're connected to the provider, the microservice (msi) will not trigger policy because msi's invoke server-side APIs. Server-side APIs do not go through the policy layer.

This explains why using `imeta` works. `imeta` invokes the client-side API (i.e. `rcModAVUMetadata()`) which goes through the policy layer.

**Comment by d-w-moore** [comment]
@dworkin  - Also, just to be clear, since this issue is in irods/irods instead of http://github.com/irods/irods_capability_indexing, what is the "desired side effect"?
   a) the creation of an AVU on the desired collection
   b) the launch of a collection operation (ie delayed task) to index objects in elasticsearch
   c) both / neither of the above

**Comment by dworkin** [comment]
> Sorry, I'll rephrase the questions.
> 
> Q. What client-side tool did you use? Q. Is the client-side tool connecting to the provider or a consumer? _(this is very important in understanding why the microservice did not run)_ Q. What client-side command did you run? Q. Are you calling `msiModAVUMetadata` from another PEP (e.g. `imkdir` -> `pep_api_coll_create`)?
> 
> Without knowing the details, I can say that if you're connected to the provider, the microservice (msi) will not trigger policy because msi's invoke server-side APIs. Server-side APIs do not go through the policy layer.
> 
> This explains why using `imeta` works. `imeta` invokes the client-side API (i.e. `rcModAVUMetadata()`) which goes through the policy layer.

The tool used was `irule`, connected to the provider, and the rule was not called through a PEP.

I understand that the metadata update for the search index is handled through the policy layer, but shouldn't it be done whenever the attribute is set, whatever the means, irrespective of the API used?

**Comment by dworkin** [comment]
> @dworkin - Also, just to be clear, since this issue is in irods/irods instead of [http://github.com/irods/irods_capability_indexing](https://github.com/irods/irods_capability_indexing), what is the "desired side effect"? a) the creation of an AVU on the desired collection b) the launch of a collection operation (ie delayed task) to index objects in elasticsearch c) both / neither of the above

I'll admit that the scope is not clear to me, and perhaps this issue should be moved to [irods_capability_indexing](https://github.com/irods/irods_capability_indexing). The "desired side effect" is b), updating the elasticsearch index.

**Comment by trel** [comment]
Oh, nice.

We weren't 100% sure remote to localhost would do it.    A workaround until we can trigger it a bit more automatically for 'in-rule' policy to fire.

**Comment by dworkin** [comment]
@d-w-moore I suspect that my setup is different: `server_config.json` has plugins for iRODS rule language, Python and elasticsearch indexing. What I see is that the rule code is executed by each plugin in turn, sometimes crashing in the indexing plugin, blocking progress even though the rule language code was run properly. Setting the INST_NAME tag has no effect.

I repeated it just now with your code fragment, log output:
```
May 27 13:11:34 pid:27872 remote addresses: 127.0.0.1, 127.0.1.1 ERROR: [exec_rule_text:687] Improperly formatted rule text in Python rule engine plugin
terminating with uncaught exception of type nlohmann::detail::parse_error: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'r'
Caught signal [6]. Dumping stacktrace and exiting
```
Followed in the log by a stack trace.

**Comment by d-w-moore** [comment]
Interesting.  What version is your iRODS server?

**Comment by dworkin** [comment]
iRODS 4.2.11, Centos 7.9

**Comment by korydraughn** [comment]
Did you pass the REP instance name via the `-r` option of `irule`?

```bash
irule -r irods_rule_engine_plugin-irods_rule_language-instance set_avu_f null null
```

The fallthrough behavior only happens if `-r` isn't passed.

**Comment by dworkin** [comment]
@korydraughn Yes I did.  The problem does not occur in how `irule` executes the rule, but in how the rule executes the remote section.

**Comment by korydraughn** [comment]
Based on this conversation, we need to do the following:
- Add support for `<INST_NAME>` to `remote()` so that it can target a specific REP
- Make sure that the plugins do not result in crashing the agent when given invalid input
- Consider the effects of allowing policy to be triggered via a microservice when there is no redirect

### Workaround
For now, you can add the metadata through an external tool (e.g. `imeta`, `PRC-based script`, etc.). You can react to PEPs using the Python REP and invoke `imeta` directly instead of using `msiModAVUMetadata()`.

Until 4.2.12 is released, you will need to either execute the command as the user who owns the collection or temporarily grant the service account permission to set metadata on the collection.

**Comment by trel** [comment]
This has come up again today for another member.     Pointed them to this discussion and current workaround.

**Comment by dworkin** [comment]
Unfortunately, this is still an issue for iRODS 4.2.12.

While it is now possible for `remote` to specify a REP, doing so will set the `irods::indexing::index` AVU without triggering the indexing event.
```
remote("localhost", "<INST_NAME>irods_rule_engine_plugin-irods_rule_language-instance</INST_NAME>") {
    msiModAVUMetadata("-C", *collection, "add", "irods::indexing::index", "yoda::metadata", "elasticsearch");
}
```
Not specifying a REP causes a crash, as before.
```
Aug  3 08:52:10 pid:21609 remote addresses: 127.0.0.1, 127.0.1.1 ERROR: [exec_rule_text:703] Improperly formatted rule text in Python rule engine plugin
terminating with uncaught exception of type nlohmann::detail::parse_error: [json.exception.parse_error.101] parse error at line 1, column 1: syntax error while parsing value - invalid literal; last read: 'r'
```

**Comment by korydraughn** [comment]
Is it safe to assume that the code block with the `remote()` call is already connected to the provider?

**Comment by dworkin** [comment]
The `remote()` call is done from server-side code; the sole purpose of the `remote()` call is to use the client API.

**Comment by korydraughn** [comment]
That's correct, but in order to trigger the client-side API, you need to be on a consumer.

If `localhost` is the provider, then the indexing event will never fire because the MSI knows it can access the database and never go through the policy layer.

If you configure the `remote` call to connect to a consumer, the indexing event should fire.

**Comment by trel** [comment]
This feels a bit cumbersome - is there a way we can make this easier / more consistent to hold?

**Comment by korydraughn** [comment]
Yes. A better solution is to expose rules through the Indexing plugin.

Just like Logical Quotas. That way, admins/users can **request** that indexing start.

**Comment by trel** [comment]
Ah yes, I like it.

**Comment by korydraughn** [comment]
That idea was mentioned back in May 2022. See https://github.com/irods/irods/issues/6422#issuecomment-1137580186.

You created an issue here: https://github.com/irods/irods_capability_indexing/issues/124

We just haven't had a chance to implement it.
